### PR TITLE
[Merged by Bors] - feat: assorted positivity extensions

### PR DESCRIPTION
**Pull request by hrmacbeth**
Positivity extensions for `NonnegHomClass` (this includes `AbsoluteValue` and `Seminorm`), `IsAbsoluteValue`, norm, the `NNReal`-to-`Real` coercion, factorials, square roots, distance (in a metric space), and diameter.

I tried to do these "properly" using Qq but I hit various errors I couldn't fix -- see 
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Qq.20doesn't.20know.20that.20two.20things.20have.20the.20same.20type
for some examples.

cc @dwrensha 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Normed/Group/Basic.lean [hunk 1]
```diff
@@ -490,21 +490,25 @@ theorem norm_nonneg' (a : E) : 0 ≤ ‖a‖ := by
...
 #align norm_nonneg norm_nonneg
 
-/- porting note: meta code, do not port
-section
+namespace Mathlib.Meta.Positivity
 
-open Tactic Tactic.Positivity
+open Lean Meta Qq Function
 
 /-- Extension for the `positivity` tactic: norms are nonnegative. -/
```
Comment (dwrensha) [inline Mathlib/Analysis/Normed/Group/Basic.lean:497]: Would be good if the comment indicated that this was just the *multiplicative* case. Maybe:
```lean
/-- Extension for the `positivity` tactic: multiplicative norms are nonnegative, via `norm_nonneg'`. -/
```

File: Mathlib/Analysis/Normed/Group/Basic.lean [hunk 2]
```diff
@@ -490,21 +490,25 @@ theorem norm_nonneg' (a : E) : 0 ≤ ‖a‖ := by
...
-#align tactic.positivity_norm tactic.positivity_norm
+@[positivity Norm.norm _]
+def evalMulNorm : PositivityExt where eval {_ _} _zα _pα e := do
+  let .app _ a ← withReducible (whnf e) | throwError "not ‖ ⬝ ‖"
+  let p ← mkAppM ``norm_nonneg' #[a]
+  pure (.nonnegative p)
 
-end
--/
+/-- Extension for the `positivity` tactic: norms are nonnegative. -/
```
Comment (dwrensha) [inline Mathlib/Analysis/Normed/Group/Basic.lean:504]: Would be good if the comment indicated that this was just the *additive* case. Maybe:
```lean
/-- Extension for the `positivity` tactic: additive norms are nonnegative, via `norm_nonneg`. -/
```

File: Mathlib/Data/Real/Sqrt.lean [hunk 3]
```diff
@@ -124,6 +124,10 @@ theorem sqrt_div (x y : ℝ≥0) : sqrt (x / y) = sqrt x / sqrt y :=
 theorem continuous_sqrt : Continuous sqrt := sqrt.continuous
 #align nnreal.continuous_sqrt NNReal.continuous_sqrt
 
+@[simp] theorem sqrt_pos : 0 < sqrt x ↔ 0 < x := by simp [pos_iff_ne_zero]
```
Comment (dwrensha) [inline Mathlib/Data/Real/Sqrt.lean:127]: It looks to me like this theorem does not exist in mathlib3. I suppose the reason we need it here and not there is that we're adding a positivity extension for `NNReal`, and no such extension exists in mathlib3?

File: test/positivity.lean [hunk 4]
```diff
@@ -279,10 +284,17 @@ example {a : ℤ} (ha : 0 < a) : (0 : ℚ) < a := by positivity
...
--- example {α : Type _} [ordered_ring α] {n : ℤ} : 0 ≤ ((n ^ 2 : ℤ) : α) := by positivity
+-- example {α : Type _} [OrderedRing α] {n : ℤ} : 0 ≤ ((n ^ 2 : ℤ) : α) := by positivity
 -- example {r : ℝ≥0} : 0 ≤ ((r : ℝ) : ereal) := by positivity
 -- example {r : ℝ≥0} : 0 < ((r + 1 : ℝ) : ereal) := by positivity
 
+/- ## Other extensions -/
+
+example [Zero β] [PartialOrder β] [NonnegHomClass F α β] (f : F) (x : α) : 0 ≤ f x := by positivity
+
+example : (0:ℝ) < ↑(3:ℝ≥0) := by positivity
```
Comment (dwrensha) [inline test/positivity.lean:295]: spaces around colons?

**Review by dwrensha — COMMENTED** [review]
I just had some nitpicks. Looks good to me.

File: Mathlib/Data/Real/Sqrt.lean (hunk 3 above)
Comment (hrmacbeth) [reply Mathlib/Data/Real/Sqrt.lean:127]: Yes, that's right!

File: Mathlib/Analysis/Normed/Group/Basic.lean (hunk 1 above)
Comment (hrmacbeth) [reply Mathlib/Analysis/Normed/Group/Basic.lean:497]:
```suggestion
/-- Extension for the `positivity` tactic: multiplicative norms are nonnegative, via `norm_nonneg'`. -/
```

File: Mathlib/Analysis/Normed/Group/Basic.lean (hunk 2 above)
Comment (hrmacbeth) [reply Mathlib/Analysis/Normed/Group/Basic.lean:504]:
```suggestion
/-- Extension for the `positivity` tactic: additive norms are nonnegative, via `norm_nonneg`. -/
```

File: test/positivity.lean (hunk 4 above)
Comment (hrmacbeth) [reply test/positivity.lean:295]:
```suggestion
example : (0 : ℝ) < ↑(3 : ℝ≥0) := by positivity
```

File: Mathlib/Analysis/Normed/Group/Basic.lean [hunk 5]
```diff
@@ -490,21 +490,25 @@ theorem norm_nonneg' (a : E) : 0 ≤ ‖a‖ := by
...
+open Lean Meta Qq Function
 
-/-- Extension for the `positivity` tactic: norms are nonnegative. -/
-@[positivity]
-unsafe def _root_.tactic.positivity_norm : expr → tactic strictness
-  | q(‖$(a)‖) =>
-    nonnegative <$> mk_app `` norm_nonneg [a] <|> nonnegative <$> mk_app `` norm_nonneg' [a]
-  | _ => failed
-#align tactic.positivity_norm tactic.positivity_norm
+/-- Extension for the `positivity` tactic: multiplicative norms are nonnegative, via `norm_nonneg'`. -/
```
Comment (hrmacbeth) [inline Mathlib/Analysis/Normed/Group/Basic.lean:497]:
```suggestion
/-- Extension for the `positivity` tactic: multiplicative norms are nonnegative, via
`norm_nonneg'`. -/
```

File: Mathlib/Algebra/Order/Hom/Basic.lean [hunk 6]
```diff
@@ -154,17 +155,19 @@ theorem le_map_div_add_map_div [Group α] [AddCommSemigroup β] [LE β] [MulLEAd
...
---   | expr.app (quote.1 ⇑(%%f)) (quote.1 (%%ₓa)) => nonnegative <$> mk_app `` map_nonneg [f, a]
---   | _ => failed
--- #align tactic.positivity_map tactic.positivity_map
+open Lean Meta Qq Function
 
---end Mathlib.Meta.Positivity
+/-- Extension for the `positivity` tactic: nonnegative maps take nonnegative values. -/
+@[positivity FunLike.coe _ _]
+def evalMap : PositivityExt where eval {_ β} _ _ e := do
+  let .app (.app _ f) a ← withReducible (whnf e)
```
[the remark this answers is not in the document]
Comment (hrmacbeth) [reply Mathlib/Algebra/Order/Hom/Basic.lean]: Thanks!

**Comment by semorrison** [comment]
bors merge

**Comment by fpvandoorn** [comment]
bors r-

I think this causes a linter failure

**Comment by fpvandoorn** [comment]
bors d+

**Pull request closed without merging**